### DISCUSSION TITLE
Remove default macOS MenuBarItems

### DISCUSSION THREAD

**Discussion by nils-kt** [dotnet/maui#11672] [Q&A]
Hello, how can I remove the default macOS MenuBarItems?
Example:
![Bildschirm­foto 2022-11-26 um 11 18 02](https://user-images.githubusercontent.com/34674720/204083906-193f63f3-a207-4e7d-b2b3-9a7bdeb6776e.png)

**Comment by headintheclouds21** [comment]
hi @nils-kt 

I was able to achieve this by overriding the BuildMenu method in the /Platforms/MacCatalyst/AppDelegate.cs file. Removing all of those menu bar items would look like

```
public override void BuildMenu(IUIMenuBuilder builder)
    {
        base.BuildMenu(builder);

        var formatMenuIdentifier = UIMenuIdentifier.Format.GetConstant();
        var helpMenuIdentifier = UIMenuIdentifier.Help.GetConstant();
        var viewMenuIdentifier = UIMenuIdentifier.View.GetConstant();
        var windowMenuIdentifier = UIMenuIdentifier.Window.GetConstant();
        var editMenuIdentifier = UIMenuIdentifier.Edit.GetConstant();
        var fileMenuIdentifier = UIMenuIdentifier.File.GetConstant();

        builder.RemoveMenu(helpMenuIdentifier);
        builder.RemoveMenu(formatMenuIdentifier);
        builder.RemoveMenu(viewMenuIdentifier);
        builder.RemoveMenu(windowMenuIdentifier);
        builder.RemoveMenu(editMenuIdentifier);
        builder.RemoveMenu(fileMenuIdentifier);
    }

**Comment by brianlellis** [comment]
Doing @headintheclouds21 solution can cause issues if you still need the menu removed. I've been able to keep the menu and remove the system default items, in Mac Catalyst, by doing so (Cannot show full code due to NDA, but should be enough to get someone started):

### Platforms/MacCatalyst/AppDelegate.cs
```csharp
public override void BuildMenu(IUIMenuBuilder builder)
{
          base.BuildMenu(builder);
          
          var fileMenuIdentifier = UIMenuIdentifier.File.GetConstant();
          UIMenu fileMenu = builder.GetMenu(fileMenuIdentifier);
                  UIMenuElement[] fileMenuChildren = fileMenu.Children;
                  builder.ReplaceMenu(UIMenuIdentifier.File.GetConstant(), fileMenu.GetMenuByReplacingChildren(new UIMenuElement[]
                  {
                      fileMenuChildren[0] // Only keep the newly designed menu in WindowMenuEvent.cs
                  }));

         // any additional code...
}
```

### WIndowMenuEvents.cs
```csharp
public WindowMenuEvents(MainPage mainPage)
{
        // Create MenuBarItems
        MenuBarItem fileMenu = new MenuBarItem { Text = "File" };

        // Create MenuFlyoutItems for File menu
        MenuFlyoutItem menuItemNew = new MenuFlyoutItem { Text = "New" };
        // openItem.Clicked += OpenFile_Clicked;
        MenuFlyoutItem menuItemOpen = new MenuFlyoutItem { Text = "Open" };
     
        MenuFlyoutItem exitItem = new MenuFlyoutItem { Text = "Quit" };
   
        // Add MenuFlyoutItems to File menu
        fileMenu.Add(menuItemNew);
        fileMenu.Add(menuItemOpen);
     
        fileMenu.Add(exitItem);
}
```

It looks like anything you created additional to the system OS menu items will be at the beginning and then the system default items are pushed onto that array.

**Comment by MSicc** [comment]
> Hello, how can I remove the default macOS MenuBarItems?
> 
> Example:
> 
> ![Bildschirm­foto 2022-11-26 um 11 18 02](https://user-images.githubusercontent.com/34674720/204083906-193f63f3-a207-4e7d-b2b3-9a7bdeb6776e.png)
> 
> 

Hi, 

I wrote a blog post on handling the MacOS menu bar: [https://msicc.net/a-deeper-look-into-menu-bar-handling-in-maccatalyst-apps-with-net-maui/](https://msicc.net/a-deeper-look-into-menu-bar-handling-in-maccatalyst-apps-with-net-maui/)

Hope it helps.
